File: phpmailer-go/oauth.go
package phpmailer

import "encoding/base64"

type OAuthTokenProvider interface {
	GetOAuth64() string
}

type SimpleOAuthTokenProvider struct {
	UserName    string
	AccessToken string
}

func NewSimpleOAuthTokenProvider(userName string, accessToken string) *SimpleOAuthTokenProvider {
	return &SimpleOAuthTokenProvider{UserName: userName, AccessToken: accessToken}
}

func (s *SimpleOAuthTokenProvider) GetOAuth64() string {
	payload := "user=" + s.UserName + "\x01auth=Bearer " + s.AccessToken + "\x01\x01"
	return base64.StdEncoding.EncodeToString([]byte(payload))
}

File: phpmailer-rs/src/oauth.rs
/// Equivalent of PHPMailer's OAuthTokenProvider interface.
pub trait OAuthTokenProvider {
    fn get_oauth64(&self) -> String;
}

#[derive(Debug, Clone)]
pub struct SimpleOAuthTokenProvider {
    user_name: String,
    access_token: String,
}

impl SimpleOAuthTokenProvider {
    pub fn new(user_name: String, access_token: String) -> Self {
        Self {
            user_name: user_name.clone(),
            access_token: access_token.clone(),
        }
    }

    pub fn user_name(&self) -> String {
        self.user_name.clone()
    }

    pub fn access_token(&self) -> String {
        self.access_token.clone()
    }
}

impl OAuthTokenProvider for SimpleOAuthTokenProvider {
    fn get_oauth64(&self) -> String {
        let auth_string = format!(
            "user={}\u{1}auth=Bearer {}\u{1}\u{1}",
            self.user_name.clone(),
            self.access_token.clone()
        );
        crate::encoding::encode_base64(auth_string.into_bytes())
    }
}
